src/driver/amdxdna/ve2_hwctx.c

```diff
@@ -571,6 +672,39 @@ int ve2_cmd_wait(struct amdxdna_ctx *hwctx, u64 seq, u32 timeout)
 	return ret > 0 ? 0 : ret;
 }
 
+void ve2_free_firmware_slots(struct amdxdna_dev_hdl *xdna_hdl, u32 max_cols)
+{
+	u32 col;
+
+	for (col = 0; col < max_cols; col++) {
+		kfree(xdna_hdl->fw_slots[col]);
+		xdna_hdl->fw_slots[col] = NULL;
+	}
+}
+
+static void timeout_cb(struct timer_list *t)
+{
+	struct amdxdna_ctx_priv *priv = from_timer(priv, t, event_timer);
```

Comment:
from_timer is removed from kernel 6.16. 
Please see: https://github.com/amd/xdna-driver/blob/main/src/driver/amdxdna/amdxdna_tdr.c#L33